// @patternfly/react-charts/dist/esm/components/Chart/Chart.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import * as React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { VictoryChart, VictoryZoomContainer } from 'victory';
import { ChartContainer } from '../ChartContainer';
import { ChartLegend, ChartLegendPosition, ChartLegendWrapper } from "../ChartLegend";
import { ChartCommonStyles } from '../ChartTheme';
import { getPaddingForSide, getTheme } from '../ChartUtils';
/**
 * See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/victory/index.d.ts
 */

export const Chart = (_ref) => {
  let {
    allowZoom = false,
    ariaDesc,
    ariaTitle,
    children,
    containerComponent = allowZoom ? React.createElement(VictoryZoomContainer, null) : React.createElement(ChartContainer, null),
    legendComponent = React.createElement(ChartLegend, null),
    legendData,
    legendPosition = ChartCommonStyles.legend.position,
    padding,
    standalone = true,
    themeColor,
    themeVariant,
    // destructure last
    theme = getTheme(themeColor, themeVariant),
    legendOrientation = theme.legend.orientation,
    height = theme.chart.height,
    width = theme.chart.width
  } = _ref,
      rest = _objectWithoutProperties(_ref, ["allowZoom", "ariaDesc", "ariaTitle", "children", "containerComponent", "legendComponent", "legendData", "legendPosition", "padding", "standalone", "themeColor", "themeVariant", "theme", "legendOrientation", "height", "width"]);

  const defaultPadding = {
    bottom: getPaddingForSide('bottom', padding, theme.chart.padding),
    left: getPaddingForSide('left', padding, theme.chart.padding),
    right: getPaddingForSide('right', padding, theme.chart.padding),
    top: getPaddingForSide('top', padding, theme.chart.padding)
  };
  const chartSize = {
    height: Math.abs(height - (defaultPadding.bottom + defaultPadding.top)),
    width: Math.abs(width - (defaultPadding.left + defaultPadding.right))
  };
  const container = React.cloneElement(containerComponent, _objectSpread({
    desc: ariaDesc,
    title: ariaTitle,
    theme
  }, containerComponent.props));
  const legend = React.cloneElement(legendComponent, _objectSpread({
    data: legendData,
    orientation: legendOrientation,
    theme
  }, legendComponent.props)); // Returns a wrapped legend

  const getWrappedLegend = () => {
    if (!legend.props.data) {
      return null;
    }

    let dx = 0;
    let dy = defaultPadding.top;

    if (legendPosition === ChartLegendPosition.bottom) {
      dy += ChartCommonStyles.legend.margin;
    } else if (legendPosition === ChartLegendPosition.bottomLeft) {
      dy += ChartCommonStyles.legend.margin;
      dx += defaultPadding.left > 10 ? defaultPadding.left - 10 : -10;
    } else if (legendPosition === ChartLegendPosition.right) {
      dx += defaultPadding.left;
    }

    return React.createElement(ChartLegendWrapper, {
      chartHeight: chartSize.height,
      chartType: "pie",
      chartWidth: chartSize.width,
      dx: dx,
      dy: dy,
      orientation: legendOrientation,
      position: legendPosition,
      svgHeight: height,
      svgWidth: width,
      theme: theme
    }, legend);
  };

  return React.createElement(VictoryChart, _extends({
    containerComponent: container,
    height: height,
    padding: defaultPadding,
    theme: theme,
    width: width
  }, rest), children, getWrappedLegend());
};
Chart.propTypes = {
  ' ': _pt.any,
  allowZoom: _pt.bool,
  animate: _pt.any,
  ariaDesc: _pt.string,
  ariaTitle: _pt.string,
  children: _pt.node,
  containerComponent: _pt.element,
  domain: _pt.any,
  domainPadding: _pt.any,
  endAngle: _pt.number,
  eventKey: _pt.any,
  events: _pt.arrayOf(_pt.any),
  externalEventMutations: _pt.arrayOf(_pt.any),
  groupComponent: _pt.element,
  height: _pt.number,
  horizontal: _pt.bool,
  innerRadius: _pt.number,
  legendComponent: _pt.element,
  legendData: _pt.arrayOf(_pt.any),
  legendOrientation: _pt.oneOf(['horizontal', 'vertical']),
  legendPosition: _pt.oneOf(['bottom', 'bottom-left', 'right']),
  maxDomain: _pt.oneOfType([_pt.number, _pt.shape({
    x: _pt.number,
    y: _pt.number
  })]),
  minDomain: _pt.oneOfType([_pt.number, _pt.shape({
    x: _pt.number,
    y: _pt.number
  })]),
  padding: _pt.any,
  polar: _pt.bool
};
hoistNonReactStatics(Chart, VictoryChart);
