// @patternfly/react-charts/dist/js/components/ChartBar/ChartBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartBar = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _hoistNonReactStatics = _interopRequireDefault(require("hoist-non-react-statics"));

var _victory = require("victory");

var _ChartUtils = require("../ChartUtils");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var ChartBar = function ChartBar(_ref) {
  var themeColor = _ref.themeColor,
      themeVariant = _ref.themeVariant,
      _ref$theme = _ref.theme,
      theme = _ref$theme === void 0 ? (0, _ChartUtils.getTheme)(themeColor, themeVariant) : _ref$theme,
      rest = _objectWithoutProperties(_ref, ["themeColor", "themeVariant", "theme"]);

  return React.createElement(_victory.VictoryBar, _extends({
    theme: theme
  }, rest));
}; // Note: VictoryBar.getDomain & VictoryBar.role must be hoisted


exports.ChartBar = ChartBar;
ChartBar.propTypes = {
  ' ': _propTypes["default"].any,
  alignment: _propTypes["default"].oneOf(["start", "middle", "end"]),
  animate: _propTypes["default"].any,
  barRatio: _propTypes["default"].number,
  barWidth: _propTypes["default"].any,
  categories: _propTypes["default"].any,
  containerComponent: _propTypes["default"].element,
  cornerRadius: _propTypes["default"].oneOfType([_propTypes["default"].any, _propTypes["default"].shape({
    top: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].any]),
    topLeft: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].any]),
    topRight: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].any]),
    bottom: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].any]),
    bottomLeft: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].any]),
    bottomRight: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].any])
  })]),
  data: _propTypes["default"].arrayOf(_propTypes["default"].any),
  dataComponent: _propTypes["default"].element,
  domain: _propTypes["default"].any,
  domainPadding: _propTypes["default"].any,
  eventKey: _propTypes["default"].any,
  events: _propTypes["default"].arrayOf(_propTypes["default"].any),
  externalEventMutations: _propTypes["default"].arrayOf(_propTypes["default"].any),
  groupComponent: _propTypes["default"].element,
  height: _propTypes["default"].number,
  horizontal: _propTypes["default"].bool,
  labelComponent: _propTypes["default"].element,
  labels: _propTypes["default"].oneOfType([_propTypes["default"].arrayOf(_propTypes["default"].string), _propTypes["default"].func]),
  maxDomain: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].shape({
    x: _propTypes["default"].number,
    y: _propTypes["default"].number
  })]),
  minDomain: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].shape({
    x: _propTypes["default"].number,
    y: _propTypes["default"].number
  })]),
  name: _propTypes["default"].string,
  origin: _propTypes["default"].shape({
    x: _propTypes["default"].number.isRequired,
    y: _propTypes["default"].number.isRequired
  }),
  padding: _propTypes["default"].any
};
(0, _hoistNonReactStatics["default"])(ChartBar, _victory.VictoryBar);
