// @patternfly/react-charts/dist/umd/components/ChartUtils/chart-origin.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports);
    global.undefined = mod.exports;
  }
})(this, function (exports) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  // Returns origin x and y coordinates
  const getChartOrigin = exports.getChartOrigin = ({
    chartHeight,
    chartWidth,
    dx = 0,
    dy = 0,
    legendPosition,
    svgWidth
  }) => ({
    x: getChartOriginX({
      chartWidth,
      dx,
      legendPosition,
      svgWidth
    }),
    y: getChartOriginY({
      chartHeight,
      dy,
      legendPosition
    })
  }); // Returns origin x coordinate


  const getChartOriginX = exports.getChartOriginX = ({
    chartWidth,
    dx = 0,
    legendPosition,
    svgWidth
  }) => {
    switch (legendPosition) {
      case 'bottom':
        return Math.round(svgWidth / 2) + dx;

      default:
        return Math.round(chartWidth / 2) + dx;
    }
  }; // Returns origin y coordinate


  const getChartOriginY = exports.getChartOriginY = ({
    chartHeight,
    dy = 0,
    legendPosition
  }) => {
    switch (legendPosition) {
      default:
        return Math.round(chartHeight / 2) + dy;
    }
  };
});
