// @patternfly/react-charts/dist/umd/components/ChartUtils/chart-theme.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "lodash", "../ChartTheme/themes/dark/blue-color-theme", "../ChartTheme/themes/dark/cyan-color-theme", "../ChartTheme/themes/dark/gold-color-theme", "../ChartTheme/themes/dark/gray-color-theme", "../ChartTheme/themes/dark/green-color-theme", "../ChartTheme/themes/dark/multi-color-ordered-theme", "../ChartTheme/themes/dark/multi-color-unordered-theme", "../ChartTheme/themes/dark/orange-color-theme", "../ChartTheme/themes/dark/purple-color-theme", "../ChartTheme/themes/light/blue-color-theme", "../ChartTheme/themes/light/cyan-color-theme", "../ChartTheme/themes/light/gold-color-theme", "../ChartTheme/themes/light/gray-color-theme", "../ChartTheme/themes/light/green-color-theme", "../ChartTheme/themes/light/multi-color-ordered-theme", "../ChartTheme/themes/light/multi-color-unordered-theme", "../ChartTheme/themes/light/orange-color-theme", "../ChartTheme/themes/light/purple-color-theme", "../ChartTheme/ChartTheme"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("lodash"), require("../ChartTheme/themes/dark/blue-color-theme"), require("../ChartTheme/themes/dark/cyan-color-theme"), require("../ChartTheme/themes/dark/gold-color-theme"), require("../ChartTheme/themes/dark/gray-color-theme"), require("../ChartTheme/themes/dark/green-color-theme"), require("../ChartTheme/themes/dark/multi-color-ordered-theme"), require("../ChartTheme/themes/dark/multi-color-unordered-theme"), require("../ChartTheme/themes/dark/orange-color-theme"), require("../ChartTheme/themes/dark/purple-color-theme"), require("../ChartTheme/themes/light/blue-color-theme"), require("../ChartTheme/themes/light/cyan-color-theme"), require("../ChartTheme/themes/light/gold-color-theme"), require("../ChartTheme/themes/light/gray-color-theme"), require("../ChartTheme/themes/light/green-color-theme"), require("../ChartTheme/themes/light/multi-color-ordered-theme"), require("../ChartTheme/themes/light/multi-color-unordered-theme"), require("../ChartTheme/themes/light/orange-color-theme"), require("../ChartTheme/themes/light/purple-color-theme"), require("../ChartTheme/ChartTheme"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.lodash, global.blueColorTheme, global.cyanColorTheme, global.goldColorTheme, global.grayColorTheme, global.greenColorTheme, global.multiColorOrderedTheme, global.multiColorUnorderedTheme, global.orangeColorTheme, global.purpleColorTheme, global.blueColorTheme, global.cyanColorTheme, global.goldColorTheme, global.grayColorTheme, global.greenColorTheme, global.multiColorOrderedTheme, global.multiColorUnorderedTheme, global.orangeColorTheme, global.purpleColorTheme, global.ChartTheme);
    global.undefined = mod.exports;
  }
})(this, function (exports, _lodash, _blueColorTheme, _cyanColorTheme, _goldColorTheme, _grayColorTheme, _greenColorTheme, _multiColorOrderedTheme, _multiColorUnorderedTheme, _orangeColorTheme, _purpleColorTheme, _blueColorTheme2, _cyanColorTheme2, _goldColorTheme2, _grayColorTheme2, _greenColorTheme2, _multiColorOrderedTheme2, _multiColorUnorderedTheme2, _orangeColorTheme2, _purpleColorTheme2, _ChartTheme) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.getTheme = exports.getLightThemeColors = exports.getDarkThemeColors = exports.getDonutUtilizationTheme = exports.getDonutThresholdStaticTheme = exports.getDonutThresholdDynamicTheme = exports.getDonutTheme = exports.getAxisTheme = exports.getCustomTheme = undefined;

  function _objectSpread(target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i] != null ? arguments[i] : {};
      var ownKeys = Object.keys(source);

      if (typeof Object.getOwnPropertySymbols === 'function') {
        ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) {
          return Object.getOwnPropertyDescriptor(source, sym).enumerable;
        }));
      }

      ownKeys.forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    }

    return target;
  }

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  // Apply custom properties to color and base themes
  const getCustomTheme = exports.getCustomTheme = (themeColor, themeVariant, customTheme) => (0, _lodash.merge)(getTheme(themeColor, themeVariant), customTheme); // Apply axis threshold properties onto base theme


  const getAxisTheme = exports.getAxisTheme = (themeColor, themeVariant) => {
    const theme = getCustomTheme(themeColor, themeVariant, _ChartTheme.ChartAxisTheme);
    return theme;
  }; // Apply donut properties onto pie chart theme


  const getDonutTheme = exports.getDonutTheme = (themeColor, themeVariant) => getCustomTheme(themeColor, themeVariant, _ChartTheme.ChartDonutTheme); // Apply dynamic donut threshold properties onto pie chart theme


  const getDonutThresholdDynamicTheme = exports.getDonutThresholdDynamicTheme = (themeColor, themeVariant) => {
    const theme = getCustomTheme(themeColor, themeVariant, _ChartTheme.ChartDonutThresholdDynamicTheme); // Merge just the first color of dynamic (blue, green, etc.) with static (grey) for expected colorScale

    theme.legend.colorScale = [theme.pie.colorScale[0], ..._ChartTheme.ChartDonutThresholdDynamicTheme.legend.colorScale]; // Merge the threshold colors in case users want to show the unused data

    theme.pie.colorScale = [theme.pie.colorScale[0], ..._ChartTheme.ChartDonutThresholdStaticTheme.pie.colorScale];
    return theme;
  }; // Apply static donut threshold properties onto pie chart theme


  const getDonutThresholdStaticTheme = exports.getDonutThresholdStaticTheme = (themeColor, themeVariant, invert) => {
    const staticTheme = (0, _lodash.cloneDeep)(_ChartTheme.ChartDonutThresholdStaticTheme);

    if (invert) {
      staticTheme.pie.colorScale = staticTheme.pie.colorScale.reverse();
    }

    return getCustomTheme(themeColor, themeVariant, staticTheme);
  }; // Apply donut utilization properties onto pie chart theme


  const getDonutUtilizationTheme = exports.getDonutUtilizationTheme = (themeColor, themeVariant) => {
    const theme = getCustomTheme(themeColor, themeVariant, _ChartTheme.ChartDonutUtilizationDynamicTheme); // Merge just the first color of dynamic (blue, green, etc.) with static (grey) for expected colorScale

    theme.pie.colorScale = [theme.pie.colorScale[0], ..._ChartTheme.ChartDonutUtilizationStaticTheme.pie.colorScale];
    theme.legend.colorScale = [theme.legend.colorScale[0], ..._ChartTheme.ChartDonutUtilizationStaticTheme.legend.colorScale];
    return theme;
  }; // Returns dark theme colors


  const getDarkThemeColors = exports.getDarkThemeColors = themeColor => {
    switch (themeColor) {
      case _ChartTheme.ChartThemeColor.blue:
        return _blueColorTheme.DarkBlueColorTheme;

      case _ChartTheme.ChartThemeColor.cyan:
        return _cyanColorTheme.DarkCyanColorTheme;

      case _ChartTheme.ChartThemeColor.gold:
        return _goldColorTheme.DarkGoldColorTheme;

      case _ChartTheme.ChartThemeColor.gray:
        return _grayColorTheme.DarkGrayColorTheme;

      case _ChartTheme.ChartThemeColor.green:
        return _greenColorTheme.DarkGreenColorTheme;

      case _ChartTheme.ChartThemeColor.multi:
      case _ChartTheme.ChartThemeColor.multiOrdered:
        return _multiColorOrderedTheme.DarkMultiColorOrderedTheme;

      case _ChartTheme.ChartThemeColor.multiUnordered:
        return _multiColorUnorderedTheme.DarkMultiColorUnorderedTheme;

      case _ChartTheme.ChartThemeColor.orange:
        return _orangeColorTheme.DarkOrangeColorTheme;

      case _ChartTheme.ChartThemeColor.purple:
        return _purpleColorTheme.DarkPurpleColorTheme;

      default:
        return _blueColorTheme.DarkBlueColorTheme;
    }
  }; // Returns light theme colors


  const getLightThemeColors = exports.getLightThemeColors = themeColor => {
    switch (themeColor) {
      case _ChartTheme.ChartThemeColor.blue:
        return _blueColorTheme2.LightBlueColorTheme;

      case _ChartTheme.ChartThemeColor.cyan:
        return _cyanColorTheme2.LightCyanColorTheme;

      case _ChartTheme.ChartThemeColor.gold:
        return _goldColorTheme2.LightGoldColorTheme;

      case _ChartTheme.ChartThemeColor.gray:
        return _grayColorTheme2.LightGrayColorTheme;

      case _ChartTheme.ChartThemeColor.green:
        return _greenColorTheme2.LightGreenColorTheme;

      case _ChartTheme.ChartThemeColor.multi:
      case _ChartTheme.ChartThemeColor.multiOrdered:
        return _multiColorOrderedTheme2.LightMultiColorOrderedTheme;

      case _ChartTheme.ChartThemeColor.multiUnordered:
        return _multiColorUnorderedTheme2.LightMultiColorUnorderedTheme;

      case _ChartTheme.ChartThemeColor.orange:
        return _orangeColorTheme2.LightOrangeColorTheme;

      case _ChartTheme.ChartThemeColor.purple:
        return _purpleColorTheme2.LightPurpleColorTheme;

      default:
        return _blueColorTheme2.LightBlueColorTheme;
    }
  }; // Applies theme color and variant to base theme


  const getTheme = exports.getTheme = (themeColor, themeVariant) => {
    // Deep clone
    const baseTheme = _objectSpread({}, JSON.parse(JSON.stringify(_ChartTheme.ChartBaseTheme)));

    switch (themeVariant) {
      case _ChartTheme.ChartThemeVariant.dark:
        return (0, _lodash.merge)(baseTheme, getDarkThemeColors(themeColor));

      case _ChartTheme.ChartThemeVariant.light:
        return (0, _lodash.merge)(baseTheme, getLightThemeColors(themeColor));

      default:
        return (0, _lodash.merge)(baseTheme, getLightThemeColors(themeColor));
    }
  };
});
