// @patternfly/react-charts/dist/umd/components/ChartDonut/ChartDonut.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "prop-types", "react", "hoist-non-react-statics", "victory", "../ChartUtils/chart-theme", "../ChartContainer", "../ChartLabel", "../ChartPie", "../ChartTheme", "../ChartUtils"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("prop-types"), require("react"), require("hoist-non-react-statics"), require("victory"), require("../ChartUtils/chart-theme"), require("../ChartContainer"), require("../ChartLabel"), require("../ChartPie"), require("../ChartTheme"), require("../ChartUtils"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.propTypes, global.react, global.hoistNonReactStatics, global.victory, global.chartTheme, global.ChartContainer, global.ChartLabel, global.ChartPie, global.ChartTheme, global.ChartUtils);
    global.undefined = mod.exports;
  }
})(this, function (exports, _propTypes, _react, _hoistNonReactStatics, _victory, _chartTheme, _ChartContainer, _ChartLabel, _ChartPie, _ChartTheme, _ChartUtils) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.ChartDonut = exports.ChartDonutSubTitlePosition = exports.ChartDonutSortOrder = exports.ChartDonutLabelPosition = undefined;

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var React = _interopRequireWildcard(_react);

  var _hoistNonReactStatics2 = _interopRequireDefault(_hoistNonReactStatics);

  function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
      return obj;
    } else {
      var newObj = {};

      if (obj != null) {
        for (var key in obj) {
          if (Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};

            if (desc.get || desc.set) {
              Object.defineProperty(newObj, key, desc);
            } else {
              newObj[key] = obj[key];
            }
          }
        }
      }

      newObj.default = obj;
      return newObj;
    }
  }

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _extends() {
    _extends = Object.assign || function (target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];

        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }

      return target;
    };

    return _extends.apply(this, arguments);
  }

  function _objectSpread(target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i] != null ? arguments[i] : {};
      var ownKeys = Object.keys(source);

      if (typeof Object.getOwnPropertySymbols === 'function') {
        ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) {
          return Object.getOwnPropertyDescriptor(source, sym).enumerable;
        }));
      }

      ownKeys.forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    }

    return target;
  }

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  function _objectWithoutProperties(source, excluded) {
    if (source == null) return {};

    var target = _objectWithoutPropertiesLoose(source, excluded);

    var key, i;

    if (Object.getOwnPropertySymbols) {
      var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

      for (i = 0; i < sourceSymbolKeys.length; i++) {
        key = sourceSymbolKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
        target[key] = source[key];
      }
    }

    return target;
  }

  function _objectWithoutPropertiesLoose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;

    for (i = 0; i < sourceKeys.length; i++) {
      key = sourceKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      target[key] = source[key];
    }

    return target;
  }

  let ChartDonutLabelPosition = exports.ChartDonutLabelPosition = undefined;

  (function (ChartDonutLabelPosition) {
    ChartDonutLabelPosition["centroid"] = "centroid";
    ChartDonutLabelPosition["endAngle"] = "endAngle";
    ChartDonutLabelPosition["startAngle"] = "startAngle";
  })(ChartDonutLabelPosition || (exports.ChartDonutLabelPosition = ChartDonutLabelPosition = {}));

  ;
  let ChartDonutSortOrder = exports.ChartDonutSortOrder = undefined;

  (function (ChartDonutSortOrder) {
    ChartDonutSortOrder["ascending"] = "ascending";
    ChartDonutSortOrder["descending"] = "descending";
  })(ChartDonutSortOrder || (exports.ChartDonutSortOrder = ChartDonutSortOrder = {}));

  ;
  let ChartDonutSubTitlePosition = exports.ChartDonutSubTitlePosition = undefined;
  /**
   * See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/victory/index.d.ts
   */

  (function (ChartDonutSubTitlePosition) {
    ChartDonutSubTitlePosition["bottom"] = "bottom";
    ChartDonutSubTitlePosition["center"] = "center";
    ChartDonutSubTitlePosition["right"] = "right";
  })(ChartDonutSubTitlePosition || (exports.ChartDonutSubTitlePosition = ChartDonutSubTitlePosition = {}));

  const ChartDonut = exports.ChartDonut = _ref => {
    let {
      ariaDesc,
      ariaTitle,
      donutDx = 0,
      donutDy = 0,
      legendPosition = _ChartTheme.ChartCommonStyles.legend.position,
      standalone = true,
      subTitle,
      subTitleComponent = React.createElement(_ChartLabel.ChartLabel, null),
      subTitleDx = 0,
      subTitleDy = 0,
      subTitlePosition = _ChartTheme.ChartDonutStyles.label.subTitlePosition,
      themeColor,
      themeVariant,
      title,
      titleComponent = React.createElement(_ChartLabel.ChartLabel, null),
      // destructure last
      theme = (0, _chartTheme.getDonutTheme)(themeColor, themeVariant),
      capHeight = 1.1,
      height = theme.pie.height,
      width = theme.pie.width,
      donutHeight = Math.min(height, width),
      donutWidth = Math.min(height, width),
      innerRadius = (Math.min(donutHeight, donutWidth) - 34) / 2
    } = _ref,
        rest = _objectWithoutProperties(_ref, ["ariaDesc", "ariaTitle", "donutDx", "donutDy", "legendPosition", "standalone", "subTitle", "subTitleComponent", "subTitleDx", "subTitleDy", "subTitlePosition", "themeColor", "themeVariant", "title", "titleComponent", "theme", "capHeight", "height", "width", "donutHeight", "donutWidth", "innerRadius"]);

    const donutSize = Math.min(donutHeight, donutWidth); // Returns subtitle

    const getSubTitle = () => {
      if (!subTitle || subTitlePosition === ChartDonutSubTitlePosition.center) {
        return null;
      }

      const subTitleProps = subTitleComponent.props ? subTitleComponent.props : {};
      return React.cloneElement(subTitleComponent, _objectSpread({
        style: _ChartTheme.ChartDonutStyles.label.subTitle,
        text: subTitle,
        textAnchor: subTitlePosition === 'right' ? 'start' : 'middle',
        verticalAnchor: 'middle',
        x: (0, _ChartUtils.getLabelX)({
          chartWidth: donutSize,
          dx: subTitleDx,
          labelPosition: subTitlePosition,
          legendPosition,
          svgWidth: width
        }),
        y: (0, _ChartUtils.getLabelY)({
          chartHeight: donutSize,
          dy: subTitleDy,
          labelPosition: subTitlePosition
        })
      }, subTitleProps));
    }; // Returns title


    const getTitle = () => {
      if (!title) {
        return null;
      }

      const titleProps = titleComponent ? titleComponent.props : {};
      const showBoth = title && subTitle && subTitlePosition == ChartDonutSubTitlePosition.center;
      return React.cloneElement(titleComponent, _objectSpread({}, showBoth && {
        capHeight
      }, {
        style: [_ChartTheme.ChartDonutStyles.label.title, _ChartTheme.ChartDonutStyles.label.subTitle],
        text: showBoth ? [title, subTitle] : title,
        textAnchor: 'middle',
        verticalAnchor: 'middle',
        x: (0, _ChartUtils.getLabelX)({
          chartWidth: donutSize,
          dx: donutDx,
          labelPosition: 'center',
          legendPosition,
          svgWidth: width
        }),
        y: (0, _ChartUtils.getLabelY)({
          chartHeight: donutSize,
          dy: donutDy,
          labelPosition: 'center'
        })
      }, titleProps));
    };

    const chart = React.createElement(_ChartPie.ChartPie, _extends({
      height: height,
      innerRadius: innerRadius > 0 ? innerRadius : 0,
      legendPosition: legendPosition,
      pieDx: donutDx,
      pieDy: donutDy,
      pieHeight: donutSize,
      pieWidth: donutSize,
      standalone: false,
      theme: theme,
      width: width
    }, rest));
    return standalone ? React.createElement(_ChartContainer.ChartContainer, {
      desc: ariaDesc,
      height: height,
      title: ariaTitle,
      width: width
    }, chart, getTitle(), getSubTitle()) : React.createElement(React.Fragment, null, chart, getTitle(), getSubTitle());
  }; // Note: VictoryPie.role must be hoisted


  ChartDonut.propTypes = {
    ' ': _propTypes2.default.any,
    animate: _propTypes2.default.any,
    ariaDesc: _propTypes2.default.string,
    ariaTitle: _propTypes2.default.string,
    capHeight: _propTypes2.default.any,
    categories: _propTypes2.default.any,
    colorScale: _propTypes2.default.any,
    containerComponent: _propTypes2.default.element,
    cornerRadius: _propTypes2.default.number,
    data: _propTypes2.default.arrayOf(_propTypes2.default.any),
    dataComponent: _propTypes2.default.element,
    donutHeight: _propTypes2.default.number,
    donutDx: _propTypes2.default.number,
    donutDy: _propTypes2.default.number,
    donutWidth: _propTypes2.default.number,
    endAngle: _propTypes2.default.number,
    eventKey: _propTypes2.default.any,
    events: _propTypes2.default.arrayOf(_propTypes2.default.any),
    externalEventMutations: _propTypes2.default.arrayOf(_propTypes2.default.any),
    groupComponent: _propTypes2.default.element,
    height: _propTypes2.default.number,
    innerRadius: _propTypes2.default.number,
    labelComponent: _propTypes2.default.element,
    labelPosition: _propTypes2.default.oneOf(['startAngle', 'endAngle', 'centroid']),
    labelRadius: _propTypes2.default.number
  };
  (0, _hoistNonReactStatics2.default)(ChartDonut, _victory.VictoryPie);
});
