// @patternfly/react-charts/dist/js/components/ChartTooltip/ChartTooltip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartTooltip = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _hoistNonReactStatics = _interopRequireDefault(require("hoist-non-react-statics"));

var _victory = require("victory");

var _ChartUtils = require("../ChartUtils");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var ChartTooltip = function ChartTooltip(_ref) {
  var themeColor = _ref.themeColor,
      themeVariant = _ref.themeVariant,
      _ref$theme = _ref.theme,
      theme = _ref$theme === void 0 ? (0, _ChartUtils.getTheme)(themeColor, themeVariant) : _ref$theme,
      rest = _objectWithoutProperties(_ref, ["themeColor", "themeVariant", "theme"]);

  return React.createElement(_victory.VictoryTooltip, _extends({
    theme: theme
  }, rest));
}; // Note: VictoryTooltip.defaultEvents must be hoisted


exports.ChartTooltip = ChartTooltip;
ChartTooltip.propTypes = {
  ' ': _propTypes["default"].any,
  active: _propTypes["default"].bool,
  activateData: _propTypes["default"].bool,
  angle: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].number]),
  cornerRadius: _propTypes["default"].any,
  data: _propTypes["default"].arrayOf(_propTypes["default"].any),
  datum: _propTypes["default"].object,
  dx: _propTypes["default"].any,
  dy: _propTypes["default"].any,
  events: _propTypes["default"].object,
  flyoutComponent: _propTypes["default"].element,
  flyoutStyle: _propTypes["default"].any,
  groupComponent: _propTypes["default"].element,
  height: _propTypes["default"].any,
  horizontal: _propTypes["default"].bool,
  index: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
  labelComponent: _propTypes["default"].element,
  orientation: _propTypes["default"].any,
  pointerLength: _propTypes["default"].any,
  pointerWidth: _propTypes["default"].any,
  renderInPortal: _propTypes["default"].bool,
  style: _propTypes["default"].any,
  text: _propTypes["default"].oneOfType([_propTypes["default"].any, _propTypes["default"].arrayOf(_propTypes["default"].string), _propTypes["default"].arrayOf(_propTypes["default"].number)]),
  theme: _propTypes["default"].any,
  themeColor: _propTypes["default"].string
};
(0, _hoistNonReactStatics["default"])(ChartTooltip, _victory.VictoryTooltip);
