// @patternfly/react-table/dist/umd/components/Table/base/provider.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "prop-types", "react", "./types"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("prop-types"), require("react"), require("./types"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.propTypes, global.react, global.types);
    global.undefined = mod.exports;
  }
})(this, function (exports, _propTypes, _react, _types) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.Provider = exports.ProviderContext = undefined;

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var React = _interopRequireWildcard(_react);

  function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
      return obj;
    } else {
      var newObj = {};

      if (obj != null) {
        for (var key in obj) {
          if (Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};

            if (desc.get || desc.set) {
              Object.defineProperty(newObj, key, desc);
            } else {
              newObj[key] = obj[key];
            }
          }
        }
      }

      newObj.default = obj;
      return newObj;
    }
  }

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _objectSpread(target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i] != null ? arguments[i] : {};
      var ownKeys = Object.keys(source);

      if (typeof Object.getOwnPropertySymbols === 'function') {
        ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) {
          return Object.getOwnPropertyDescriptor(source, sym).enumerable;
        }));
      }

      ownKeys.forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    }

    return target;
  }

  function _objectWithoutProperties(source, excluded) {
    if (source == null) return {};

    var target = _objectWithoutPropertiesLoose(source, excluded);

    var key, i;

    if (Object.getOwnPropertySymbols) {
      var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

      for (i = 0; i < sourceSymbolKeys.length; i++) {
        key = sourceSymbolKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
        target[key] = source[key];
      }
    }

    return target;
  }

  function _objectWithoutPropertiesLoose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;

    for (i = 0; i < sourceKeys.length; i++) {
      key = sourceKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      target[key] = source[key];
    }

    return target;
  }

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }
  /**
   * provider.tsx
   *
   * Forked from reactabular-table version 8.14.0
   * https://github.com/reactabular/reactabular/tree/v8.14.0/packages/reactabular-table/src
   */


  const ProviderContext = exports.ProviderContext = React.createContext({
    columns: null,
    renderers: null
  });

  class Provider extends React.Component {
    render() {
      const _this$props = this.props,
            {
        columns,
        renderers,
        components,
        children
      } = _this$props,
            props = _objectWithoutProperties(_this$props, ["columns", "renderers", "components", "children"]);

      let finalRenderers = renderers;

      if (components) {
        // tslint:disable-next-line:no-console
        console.warn('`components` have been deprecated in favor of `renderers` and will be removed in the next major version, please rename!');
        finalRenderers = components;
      }

      const provider = React.createElement(renderers.table || _types.TableDefaults.renderers.table, props, children);
      return React.createElement(ProviderContext.Provider, {
        value: {
          columns,
          renderers: {
            table: finalRenderers.table || _types.TableDefaults.renderers.table,
            header: _objectSpread({}, _types.TableDefaults.renderers.header, finalRenderers.header),
            body: _objectSpread({}, _types.TableDefaults.renderers.body, finalRenderers.body)
          }
        }
      }, provider);
    }

  }

  exports.Provider = Provider;

  _defineProperty(Provider, "propTypes", {
    children: _propTypes2.default.node,
    role: _propTypes2.default.string,
    className: _propTypes2.default.string
  });

  _defineProperty(Provider, "defaultProps", {
    renderers: _types.TableDefaults.renderers
  });

  ;
});
