// @patternfly/react-charts/dist/esm/components/ChartGroup/ChartGroup.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import * as React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { VictoryGroup, VictoryZoomContainer } from 'victory';
import { getTheme } from '../ChartUtils';
import { ChartContainer } from "../ChartContainer";
export let ChartGroupSortOrder;

(function (ChartGroupSortOrder) {
  ChartGroupSortOrder["ascending"] = "ascending";
  ChartGroupSortOrder["descending"] = "descending";
})(ChartGroupSortOrder || (ChartGroupSortOrder = {}));

;
/**
 * See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/victory/index.d.ts
 */

;
export const ChartGroup = (_ref) => {
  let {
    allowZoom = false,
    ariaDesc,
    ariaTitle,
    children,
    containerComponent = allowZoom ? React.createElement(VictoryZoomContainer, null) : React.createElement(ChartContainer, null),
    themeColor,
    themeVariant,
    // destructure last
    theme = getTheme(themeColor, themeVariant)
  } = _ref,
      rest = _objectWithoutProperties(_ref, ["allowZoom", "ariaDesc", "ariaTitle", "children", "containerComponent", "themeColor", "themeVariant", "theme"]);

  const container = React.cloneElement(containerComponent, _objectSpread({
    desc: ariaDesc,
    title: ariaTitle,
    theme
  }, containerComponent.props));
  return React.createElement(VictoryGroup, _extends({
    containerComponent: container,
    theme: theme
  }, rest), children);
}; // Note: VictoryGroup.role must be hoisted

ChartGroup.propTypes = {
  ' ': _pt.any,
  allowZoom: _pt.bool,
  animate: _pt.any,
  ariaDesc: _pt.string,
  ariaTitle: _pt.string,
  categories: _pt.any,
  children: _pt.node,
  color: _pt.string,
  colorScale: _pt.any,
  containerComponent: _pt.element,
  data: _pt.arrayOf(_pt.any),
  domain: _pt.any,
  domainPadding: _pt.any,
  eventKey: _pt.any,
  events: _pt.arrayOf(_pt.any),
  externalEventMutations: _pt.arrayOf(_pt.any),
  groupComponent: _pt.element,
  height: _pt.number,
  horizontal: _pt.bool,
  labelComponent: _pt.element,
  labels: _pt.oneOfType([_pt.arrayOf(_pt.string), _pt.func]),
  maxDomain: _pt.oneOfType([_pt.number, _pt.shape({
    x: _pt.number,
    y: _pt.number
  })]),
  minDomain: _pt.oneOfType([_pt.number, _pt.shape({
    x: _pt.number,
    y: _pt.number
  })]),
  name: _pt.string,
  offset: _pt.number
};
hoistNonReactStatics(ChartGroup, VictoryGroup);
