// @patternfly/react-charts/dist/esm/components/ChartTheme/ChartTheme.js
import { AxisTheme } from './themes/axis-theme';
import { BaseTheme } from './themes/base-theme';
import { DonutTheme } from './themes/donut-theme';
import { DonutThresholdDynamicTheme, DonutThresholdStaticTheme } from './themes/donut-threshold-theme';
import { DonutUtilizationDynamicTheme, DonutUtilizationStaticTheme } from './themes/donut-utilization-theme';

/**
 * The multiOrdered theme is intended for ordered charts; donut, pie, bar, & stack
 * The multiUnordered theme is intended for unordered charts; area & line
 *
 * Note: multi defaults to multiOrdered
 */
export const ChartThemeColor = {
  blue: 'blue',
  cyan: 'cyan',
  default: 'blue',
  gold: 'gold',
  gray: 'gray',
  green: 'green',
  multi: 'multi',
  multiOrdered: 'multi-ordered',
  multiUnordered: 'multi-unordered',
  orange: 'orange',
  purple: 'purple'
};
export const ChartThemeVariant = {
  dark: 'dark',
  default: 'light',
  light: 'light'
};
export const ChartAxisTheme = AxisTheme;
export const ChartBaseTheme = BaseTheme;
export const ChartDonutUtilizationDynamicTheme = DonutUtilizationDynamicTheme;
export const ChartDonutUtilizationStaticTheme = DonutUtilizationStaticTheme;
export const ChartDonutTheme = DonutTheme;
export const ChartDonutThresholdDynamicTheme = DonutThresholdDynamicTheme;
export const ChartDonutThresholdStaticTheme = DonutThresholdStaticTheme;
