// @patternfly/react-charts/dist/esm/components/ChartPie/ChartPie.js
import _pt from "prop-types";

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import * as React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { VictoryPie } from 'victory';
import { ChartContainer } from '../ChartContainer';
import { ChartLegend, ChartLegendWrapper } from '../ChartLegend';
import { ChartCommonStyles } from '../ChartTheme';
import { ChartTooltip } from '../ChartTooltip';
import { getChartOrigin, getTheme } from '../ChartUtils';
export let ChartPieLabelPosition;

(function (ChartPieLabelPosition) {
  ChartPieLabelPosition["centroid"] = "centroid";
  ChartPieLabelPosition["endAngle"] = "endAngle";
  ChartPieLabelPosition["startAngle"] = "startAngle";
})(ChartPieLabelPosition || (ChartPieLabelPosition = {}));

;
export let ChartPieLegendPosition;

(function (ChartPieLegendPosition) {
  ChartPieLegendPosition["bottom"] = "bottom";
  ChartPieLegendPosition["right"] = "right";
})(ChartPieLegendPosition || (ChartPieLegendPosition = {}));

;
export let ChartPieSortOrder;

(function (ChartPieSortOrder) {
  ChartPieSortOrder["ascending"] = "ascending";
  ChartPieSortOrder["descending"] = "descending";
})(ChartPieSortOrder || (ChartPieSortOrder = {}));

;
/**
 * See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/victory/index.d.ts
 */

export const ChartPie = (_ref) => {
  let {
    ariaDesc,
    ariaTitle,
    pieDx = 0,
    pieDy = 0,
    legendComponent = React.createElement(ChartLegend, null),
    legendData,
    legendDx = 0,
    legendDy = 0,
    legendPosition = ChartCommonStyles.legend.position,
    standalone = true,
    themeColor,
    themeVariant,
    // destructure last
    theme = getTheme(themeColor, themeVariant),
    labelComponent = React.createElement(ChartTooltip, {
      theme: theme
    }),
    legendOrientation = theme.legend.orientation,
    height = theme.pie.height,
    width = theme.pie.width,
    pieHeight = Math.min(height, width),
    pieWidth = Math.min(height, width)
  } = _ref,
      rest = _objectWithoutProperties(_ref, ["ariaDesc", "ariaTitle", "pieDx", "pieDy", "legendComponent", "legendData", "legendDx", "legendDy", "legendPosition", "standalone", "themeColor", "themeVariant", "theme", "labelComponent", "legendOrientation", "height", "width", "pieHeight", "pieWidth"]);

  const pieSize = Math.min(pieHeight, pieWidth);
  const chart = React.createElement(VictoryPie, _extends({
    height: pieSize,
    labelComponent: labelComponent,
    origin: getChartOrigin({
      chartHeight: pieSize,
      chartWidth: pieSize,
      dx: pieDx,
      dy: pieDy,
      legendPosition,
      svgWidth: width
    }),
    standalone: false,
    theme: theme,
    width: pieSize
  }, rest));
  const legend = React.cloneElement(legendComponent, _objectSpread({
    data: legendData,
    orientation: legendOrientation,
    theme
  }, legendComponent.props)); // Returns a wrapped legend

  const getWrappedLegend = () => {
    if (!legend.props.data) {
      return null;
    }

    return React.createElement(ChartLegendWrapper, {
      chartHeight: pieSize,
      chartType: "pie",
      chartWidth: pieSize,
      dx: legendDx,
      dy: legendDy,
      orientation: legendOrientation,
      position: legendPosition,
      svgHeight: height,
      svgWidth: width,
      theme: theme
    }, legend);
  };

  return standalone ? React.createElement(ChartContainer, {
    desc: ariaDesc,
    height: height,
    title: ariaTitle,
    width: width
  }, chart, getWrappedLegend()) : React.createElement(React.Fragment, null, chart, getWrappedLegend());
}; // Note: VictoryPie.role must be hoisted

ChartPie.propTypes = {
  ' ': _pt.any,
  animate: _pt.any,
  ariaDesc: _pt.string,
  ariaTitle: _pt.string,
  categories: _pt.any,
  colorScale: _pt.any,
  containerComponent: _pt.element,
  cornerRadius: _pt.number,
  data: _pt.arrayOf(_pt.any),
  dataComponent: _pt.element,
  endAngle: _pt.number,
  eventKey: _pt.any,
  events: _pt.arrayOf(_pt.any),
  externalEventMutations: _pt.arrayOf(_pt.any),
  groupComponent: _pt.element,
  height: _pt.number,
  innerRadius: _pt.number,
  labelComponent: _pt.element,
  labelPosition: _pt.oneOf(['startAngle', 'endAngle', 'centroid']),
  labelRadius: _pt.number,
  labels: _pt.oneOfType([_pt.arrayOf(_pt.string), _pt.func]),
  legendComponent: _pt.element,
  legendData: _pt.arrayOf(_pt.any),
  legendDx: _pt.number,
  legendDy: _pt.number
};
hoistNonReactStatics(ChartPie, VictoryPie);
