// @patternfly/react-charts/dist/js/components/ChartUtils/chart-theme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getTheme = exports.getLightThemeColors = exports.getDarkThemeColors = exports.getDonutUtilizationTheme = exports.getDonutThresholdStaticTheme = exports.getDonutThresholdDynamicTheme = exports.getDonutTheme = exports.getAxisTheme = exports.getCustomTheme = void 0;

var _lodash = require("lodash");

var _blueColorTheme = require("../ChartTheme/themes/dark/blue-color-theme");

var _cyanColorTheme = require("../ChartTheme/themes/dark/cyan-color-theme");

var _goldColorTheme = require("../ChartTheme/themes/dark/gold-color-theme");

var _grayColorTheme = require("../ChartTheme/themes/dark/gray-color-theme");

var _greenColorTheme = require("../ChartTheme/themes/dark/green-color-theme");

var _multiColorOrderedTheme = require("../ChartTheme/themes/dark/multi-color-ordered-theme");

var _multiColorUnorderedTheme = require("../ChartTheme/themes/dark/multi-color-unordered-theme");

var _orangeColorTheme = require("../ChartTheme/themes/dark/orange-color-theme");

var _purpleColorTheme = require("../ChartTheme/themes/dark/purple-color-theme");

var _blueColorTheme2 = require("../ChartTheme/themes/light/blue-color-theme");

var _cyanColorTheme2 = require("../ChartTheme/themes/light/cyan-color-theme");

var _goldColorTheme2 = require("../ChartTheme/themes/light/gold-color-theme");

var _grayColorTheme2 = require("../ChartTheme/themes/light/gray-color-theme");

var _greenColorTheme2 = require("../ChartTheme/themes/light/green-color-theme");

var _multiColorOrderedTheme2 = require("../ChartTheme/themes/light/multi-color-ordered-theme");

var _multiColorUnorderedTheme2 = require("../ChartTheme/themes/light/multi-color-unordered-theme");

var _orangeColorTheme2 = require("../ChartTheme/themes/light/orange-color-theme");

var _purpleColorTheme2 = require("../ChartTheme/themes/light/purple-color-theme");

var _ChartTheme = require("../ChartTheme/ChartTheme");

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

// Apply custom properties to color and base themes
var getCustomTheme = function getCustomTheme(themeColor, themeVariant, customTheme) {
  return (0, _lodash.merge)(getTheme(themeColor, themeVariant), customTheme);
}; // Apply axis threshold properties onto base theme


exports.getCustomTheme = getCustomTheme;

var getAxisTheme = function getAxisTheme(themeColor, themeVariant) {
  var theme = getCustomTheme(themeColor, themeVariant, _ChartTheme.ChartAxisTheme);
  return theme;
}; // Apply donut properties onto pie chart theme


exports.getAxisTheme = getAxisTheme;

var getDonutTheme = function getDonutTheme(themeColor, themeVariant) {
  return getCustomTheme(themeColor, themeVariant, _ChartTheme.ChartDonutTheme);
}; // Apply dynamic donut threshold properties onto pie chart theme


exports.getDonutTheme = getDonutTheme;

var getDonutThresholdDynamicTheme = function getDonutThresholdDynamicTheme(themeColor, themeVariant) {
  var theme = getCustomTheme(themeColor, themeVariant, _ChartTheme.ChartDonutThresholdDynamicTheme); // Merge just the first color of dynamic (blue, green, etc.) with static (grey) for expected colorScale

  theme.legend.colorScale = [theme.pie.colorScale[0]].concat(_toConsumableArray(_ChartTheme.ChartDonutThresholdDynamicTheme.legend.colorScale)); // Merge the threshold colors in case users want to show the unused data

  theme.pie.colorScale = [theme.pie.colorScale[0]].concat(_toConsumableArray(_ChartTheme.ChartDonutThresholdStaticTheme.pie.colorScale));
  return theme;
}; // Apply static donut threshold properties onto pie chart theme


exports.getDonutThresholdDynamicTheme = getDonutThresholdDynamicTheme;

var getDonutThresholdStaticTheme = function getDonutThresholdStaticTheme(themeColor, themeVariant, invert) {
  var staticTheme = (0, _lodash.cloneDeep)(_ChartTheme.ChartDonutThresholdStaticTheme);

  if (invert) {
    staticTheme.pie.colorScale = staticTheme.pie.colorScale.reverse();
  }

  return getCustomTheme(themeColor, themeVariant, staticTheme);
}; // Apply donut utilization properties onto pie chart theme


exports.getDonutThresholdStaticTheme = getDonutThresholdStaticTheme;

var getDonutUtilizationTheme = function getDonutUtilizationTheme(themeColor, themeVariant) {
  var theme = getCustomTheme(themeColor, themeVariant, _ChartTheme.ChartDonutUtilizationDynamicTheme); // Merge just the first color of dynamic (blue, green, etc.) with static (grey) for expected colorScale

  theme.pie.colorScale = [theme.pie.colorScale[0]].concat(_toConsumableArray(_ChartTheme.ChartDonutUtilizationStaticTheme.pie.colorScale));
  theme.legend.colorScale = [theme.legend.colorScale[0]].concat(_toConsumableArray(_ChartTheme.ChartDonutUtilizationStaticTheme.legend.colorScale));
  return theme;
}; // Returns dark theme colors


exports.getDonutUtilizationTheme = getDonutUtilizationTheme;

var getDarkThemeColors = function getDarkThemeColors(themeColor) {
  switch (themeColor) {
    case _ChartTheme.ChartThemeColor.blue:
      return _blueColorTheme.DarkBlueColorTheme;

    case _ChartTheme.ChartThemeColor.cyan:
      return _cyanColorTheme.DarkCyanColorTheme;

    case _ChartTheme.ChartThemeColor.gold:
      return _goldColorTheme.DarkGoldColorTheme;

    case _ChartTheme.ChartThemeColor.gray:
      return _grayColorTheme.DarkGrayColorTheme;

    case _ChartTheme.ChartThemeColor.green:
      return _greenColorTheme.DarkGreenColorTheme;

    case _ChartTheme.ChartThemeColor.multi:
    case _ChartTheme.ChartThemeColor.multiOrdered:
      return _multiColorOrderedTheme.DarkMultiColorOrderedTheme;

    case _ChartTheme.ChartThemeColor.multiUnordered:
      return _multiColorUnorderedTheme.DarkMultiColorUnorderedTheme;

    case _ChartTheme.ChartThemeColor.orange:
      return _orangeColorTheme.DarkOrangeColorTheme;

    case _ChartTheme.ChartThemeColor.purple:
      return _purpleColorTheme.DarkPurpleColorTheme;

    default:
      return _blueColorTheme.DarkBlueColorTheme;
  }
}; // Returns light theme colors


exports.getDarkThemeColors = getDarkThemeColors;

var getLightThemeColors = function getLightThemeColors(themeColor) {
  switch (themeColor) {
    case _ChartTheme.ChartThemeColor.blue:
      return _blueColorTheme2.LightBlueColorTheme;

    case _ChartTheme.ChartThemeColor.cyan:
      return _cyanColorTheme2.LightCyanColorTheme;

    case _ChartTheme.ChartThemeColor.gold:
      return _goldColorTheme2.LightGoldColorTheme;

    case _ChartTheme.ChartThemeColor.gray:
      return _grayColorTheme2.LightGrayColorTheme;

    case _ChartTheme.ChartThemeColor.green:
      return _greenColorTheme2.LightGreenColorTheme;

    case _ChartTheme.ChartThemeColor.multi:
    case _ChartTheme.ChartThemeColor.multiOrdered:
      return _multiColorOrderedTheme2.LightMultiColorOrderedTheme;

    case _ChartTheme.ChartThemeColor.multiUnordered:
      return _multiColorUnorderedTheme2.LightMultiColorUnorderedTheme;

    case _ChartTheme.ChartThemeColor.orange:
      return _orangeColorTheme2.LightOrangeColorTheme;

    case _ChartTheme.ChartThemeColor.purple:
      return _purpleColorTheme2.LightPurpleColorTheme;

    default:
      return _blueColorTheme2.LightBlueColorTheme;
  }
}; // Applies theme color and variant to base theme


exports.getLightThemeColors = getLightThemeColors;

var getTheme = function getTheme(themeColor, themeVariant) {
  // Deep clone
  var baseTheme = _objectSpread({}, JSON.parse(JSON.stringify(_ChartTheme.ChartBaseTheme)));

  switch (themeVariant) {
    case _ChartTheme.ChartThemeVariant.dark:
      return (0, _lodash.merge)(baseTheme, getDarkThemeColors(themeColor));

    case _ChartTheme.ChartThemeVariant.light:
      return (0, _lodash.merge)(baseTheme, getLightThemeColors(themeColor));

    default:
      return (0, _lodash.merge)(baseTheme, getLightThemeColors(themeColor));
  }
};

exports.getTheme = getTheme;
