// @patternfly/react-charts/dist/js/components/ChartUtils/chart-origin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getChartOriginY = exports.getChartOriginX = exports.getChartOrigin = void 0;

// Returns origin x and y coordinates
var getChartOrigin = function getChartOrigin(_ref) {
  var chartHeight = _ref.chartHeight,
      chartWidth = _ref.chartWidth,
      _ref$dx = _ref.dx,
      dx = _ref$dx === void 0 ? 0 : _ref$dx,
      _ref$dy = _ref.dy,
      dy = _ref$dy === void 0 ? 0 : _ref$dy,
      legendPosition = _ref.legendPosition,
      svgWidth = _ref.svgWidth;
  return {
    x: getChartOriginX({
      chartWidth: chartWidth,
      dx: dx,
      legendPosition: legendPosition,
      svgWidth: svgWidth
    }),
    y: getChartOriginY({
      chartHeight: chartHeight,
      dy: dy,
      legendPosition: legendPosition
    })
  };
}; // Returns origin x coordinate


exports.getChartOrigin = getChartOrigin;

var getChartOriginX = function getChartOriginX(_ref2) {
  var chartWidth = _ref2.chartWidth,
      _ref2$dx = _ref2.dx,
      dx = _ref2$dx === void 0 ? 0 : _ref2$dx,
      legendPosition = _ref2.legendPosition,
      svgWidth = _ref2.svgWidth;

  switch (legendPosition) {
    case 'bottom':
      return Math.round(svgWidth / 2) + dx;

    default:
      return Math.round(chartWidth / 2) + dx;
  }
}; // Returns origin y coordinate


exports.getChartOriginX = getChartOriginX;

var getChartOriginY = function getChartOriginY(_ref3) {
  var chartHeight = _ref3.chartHeight,
      _ref3$dy = _ref3.dy,
      dy = _ref3$dy === void 0 ? 0 : _ref3$dy,
      legendPosition = _ref3.legendPosition;

  switch (legendPosition) {
    default:
      return Math.round(chartHeight / 2) + dy;
  }
};

exports.getChartOriginY = getChartOriginY;
