// @patternfly/react-charts/dist/js/components/ChartTheme/ChartTheme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartDonutThresholdStaticTheme = exports.ChartDonutThresholdDynamicTheme = exports.ChartDonutTheme = exports.ChartDonutUtilizationStaticTheme = exports.ChartDonutUtilizationDynamicTheme = exports.ChartBaseTheme = exports.ChartAxisTheme = exports.ChartThemeVariant = exports.ChartThemeColor = void 0;

var _axisTheme = require("./themes/axis-theme");

var _baseTheme = require("./themes/base-theme");

var _donutTheme = require("./themes/donut-theme");

var _donutThresholdTheme = require("./themes/donut-threshold-theme");

var _donutUtilizationTheme = require("./themes/donut-utilization-theme");

/**
 * The multiOrdered theme is intended for ordered charts; donut, pie, bar, & stack
 * The multiUnordered theme is intended for unordered charts; area & line
 *
 * Note: multi defaults to multiOrdered
 */
var ChartThemeColor = {
  blue: 'blue',
  cyan: 'cyan',
  "default": 'blue',
  gold: 'gold',
  gray: 'gray',
  green: 'green',
  multi: 'multi',
  multiOrdered: 'multi-ordered',
  multiUnordered: 'multi-unordered',
  orange: 'orange',
  purple: 'purple'
};
exports.ChartThemeColor = ChartThemeColor;
var ChartThemeVariant = {
  dark: 'dark',
  "default": 'light',
  light: 'light'
};
exports.ChartThemeVariant = ChartThemeVariant;
var ChartAxisTheme = _axisTheme.AxisTheme;
exports.ChartAxisTheme = ChartAxisTheme;
var ChartBaseTheme = _baseTheme.BaseTheme;
exports.ChartBaseTheme = ChartBaseTheme;
var ChartDonutUtilizationDynamicTheme = _donutUtilizationTheme.DonutUtilizationDynamicTheme;
exports.ChartDonutUtilizationDynamicTheme = ChartDonutUtilizationDynamicTheme;
var ChartDonutUtilizationStaticTheme = _donutUtilizationTheme.DonutUtilizationStaticTheme;
exports.ChartDonutUtilizationStaticTheme = ChartDonutUtilizationStaticTheme;
var ChartDonutTheme = _donutTheme.DonutTheme;
exports.ChartDonutTheme = ChartDonutTheme;
var ChartDonutThresholdDynamicTheme = _donutThresholdTheme.DonutThresholdDynamicTheme;
exports.ChartDonutThresholdDynamicTheme = ChartDonutThresholdDynamicTheme;
var ChartDonutThresholdStaticTheme = _donutThresholdTheme.DonutThresholdStaticTheme;
exports.ChartDonutThresholdStaticTheme = ChartDonutThresholdStaticTheme;
